test-data/unit/check-recursive-types.test

```diff
@@ -438,3 +438,230 @@ E = List[E[E[T]]]  # E: Invalid recursive alias: type variable nesting on right
+class B(NamedTuple):
+    x: A
+    y: int
+class A(NamedTuple):
+    x: str
+    y: B
+
+n: A
+s: str = n.y[0]  # E: Incompatible types in assignment (expression has type "A", variable has type "str")
+
+m: B
+n = m.x
+n = n.y.x
+
+t: Tuple[str, B]
+t = n
+t = m  # E: Incompatible types in assignment (expression has type "B", variable has type "Tuple[str, B]")
+[builtins fixtures/tuple.pyi]
+
+[case testMutuallyRecursiveNamedTuplesCalls]
+# flags: --enable-recursive-aliases
+from typing import NamedTuple
+
+B = NamedTuple('B', [('x', A), ('y', int)])
+A = NamedTuple('A', [('x', str), ('y', 'B')])
+n: A
+def f(m: B) -> None: pass
+reveal_type(n) # N: Revealed type is "Tuple[builtins.str, Tuple[..., builtins.int, fallback=__main__.B], fallback=__main__.A]"
+reveal_type(f) # N: Revealed type is "def (m: Tuple[Tuple[builtins.str, ..., fallback=__main__.A], builtins.int, fallback=__main__.B])"
+f(n)  # E: Argument 1 to "f" has incompatible type "A"; expected "B"
+[builtins fixtures/tuple.pyi]
+
+[case testNoRecursiveTuplesAtFunctionScope]
+# flags: --enable-recursive-aliases
+from typing import NamedTuple, Tuple
+def foo() -> None:
+    class B(NamedTuple):
+        x: B  # E: Cannot resolve name "B" (possible cyclic definition) \
+              # N: Recursive types are not allowed at function scope
+        y: int
+    b: B
+    reveal_type(b)  # N: Revealed type is "Tuple[Any, builtins.int, fallback=__main__.B@4]"
+[builtins fixtures/tuple.pyi]
+
+[case testBasicRecursiveTypedDictClass]
+# flags: --enable-recursive-aliases
+from typing import TypedDict
+
+class TD(TypedDict):
+    x: int
+    y: TD
+
+td: TD
+reveal_type(td)  # N: Revealed type is "TypedDict('__main__.TD', {'x': builtins.int, 'y': ...})"
+s: str = td["y"]  # E: Incompatible types in assignment (expression has type "TD", variable has type "str")
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+[case testBasicRecursiveTypedDictCall]
+# flags: --enable-recursive-aliases
+from typing import TypedDict
+
+TD = TypedDict("TD", {"x": int, "y": TD})
+td: TD
+reveal_type(td)  # N: Revealed type is "TypedDict('__main__.TD', {'x': builtins.int, 'y': ...})"
+
+TD2 = TypedDict("TD2", {"x": int, "y": TD2})
+td2: TD2
+TD3 = TypedDict("TD3", {"x": str, "y": TD3})
+td3: TD3
+
+td = td2
+td = td3  # E: Incompatible types in assignment (expression has type "TD3", variable has type "TD")
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+[case testBasicRecursiveTypedDictExtending]
+# flags: --enable-recursive-aliases
+from typing import TypedDict
+
+class TDA(TypedDict):
+    xa: int
+    ya: TD
+
+class TDB(TypedDict):
+    xb: int
+    yb: TD
+
+class TD(TDA, TDB):
+    a: TDA
+    b: TDB
+
+td: TD
+reveal_type(td)  # N: Revealed type is "TypedDict('__main__.TD', {'xb': builtins.int, 'yb': ..., 'xa': builtins.int, 'ya': ..., 'a': TypedDict('__main__.TDA', {'xa': builtins.int, 'ya': ...}), 'b': TypedDict('__main__.TDB', {'xb': builtins.int, 'yb': ...})})"
```

Comment:
Test constructing recursive TypedDicts, e.g. `TD(...)` (with different levels of nesting) and `td: TD = {...}`.

Test calling some methods on recursive typeddicts.

Test indexed assignment to a recursive typeddict.